src/jvm/main/org/jetbrains/kotlinx/lincheck/execution/ExecutionScenario.kt

```diff
@@ -0,0 +1,150 @@
+/*
+ * Lincheck
+ *
+ * Copyright (C) 2019 - 2023 JetBrains s.r.o.
+ *
+ * This Source Code Form is subject to the terms of the
+ * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
+ * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package org.jetbrains.kotlinx.lincheck.execution
+
+import org.jetbrains.kotlinx.lincheck.*
+
+/**
+ * This class represents an execution scenario.
+ *
+ * Execution scenarios can be run via [Strategy] to produce an [ExecutionResult].
+ * The [ExecutionGenerator] class is responsible for generation of random scenarios.
+ */
+class ExecutionScenario(
+    /**
+     * The initial sequential part of the execution.
+     * It helps to produce different initial states before the parallel part.
+     *
+     * The initial part is executed in the first thread of the scenario before the parallel part.
+     *
+     * The initial part should contain only non-suspendable actors;
+     * otherwise, the single initial execution thread will suspend with no chance to be resumed.
+     */
+    val initExecution: List<Actor>,
+    /**
+     * The parallel part of the execution.
+     * It is used to find an interleaving with incorrect behaviour.
+     */
+    val parallelExecution: List<List<Actor>>,
+    /**
+     * The post sequential part of the execution.
+     * It is commonly used to test that the data structure is in some valid state.
+     *
+     * The post part is executed in the first thread of the scenario after the parallel part.
+     *
+     * If this execution scenario contains suspendable actors, the post part should be empty;
+     * if not, an actor could resume a previously suspended one from the parallel execution part.
+     */
+    val postExecution: List<Actor>
+) {
+    /**
+     * Number of threads used by this execution.
+     */
+    val nThreads: Int
+        get() = parallelExecution.size
+
+    /**
+     * List containing for each thread its list of actors.
+     * Init and post parts are placed in the 1st thread.
+     */
+    val threads: List<List<Actor>> = (0 until nThreads).map { i ->
```

Comment:
You do not need to copy the scenario. More importantly, you do not need this `threads` property -- just add a `get(threadId, actorId)` function.

Reply:
This is actually not that simple :(

In some cases, I also need, for example, to get the size of the particular thread (including init/post parts for the 1st thread). By having `threads` as an array, I have all the usual API for `threads[0]`, e.g. `size`, `indices`, etc. 
Otherwise, I need to somehow duplicate this API in `ExecutionScenario` class. 

I actually tried the approach you suggest at first, but the code turned out to be more complicated. 

If the memory consumption is a concern, we can also try another approach, we can have `init`, `post` and `parallel[i]` part lists as sub-lists of `threads[i]` lists.

Reply:
Ok. Then maybe make `ExecutionScenario : List<List<Actor>> by threads`. Will it make code easier to read?

Reply:
I cannot delegate to a property not passed into constructor. 
There is a workaround with private primary constructor and public secondary constructor:
https://stackoverflow.com/a/71955281/4676150 

Do we want to use it here?

Reply:
If it makes the code easier to read, yes. You decide.